Flaky-test: MLPendingAckStoreTest.testMainProcess

### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.


### Example failure

https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706

### Exception stacktrace

```
Error:  Tests run: 16, Failures: 1, Errors: 0, Skipped: 14, Time elapsed: 19.252 s <<< FAILURE! - in org.apache.pulsar.broker.transaction.pendingack.impl.MLPendingAckStoreTest
[807](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:808)
  Error:  testMainProcess(org.apache.pulsar.broker.transaction.pendingack.impl.MLPendingAckStoreTest)  Time elapsed: 0.311 s  <<< FAILURE!
[808](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:809)
  java.util.concurrent.ExecutionException: org.apache.bookkeeper.mledger.ManagedLedgerException$BadVersionException: org.apache.pulsar.metadata.api.MetadataStoreException$BadVersionException: org.apache.zookeeper.KeeperException$BadVersionException: KeeperErrorCode = BadVersion for /managed-ledgers/tnx/ns1/persistent/test-txn-topic-test__transaction_pending_ack/__pending_ack_state
[809](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:810)
  	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
[810](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:811)
  	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
[811](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:812)
  	at org.apache.pulsar.broker.transaction.pendingack.impl.MLPendingAckStoreTest.testMainProcess(MLPendingAckStoreTest.java:259)
[812](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:813)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[813](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:814)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
[814](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:815)
  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[815](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:816)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
[816](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:817)
  	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
[817](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:818)
  	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
[818](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:819)
  	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
[819](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:820)
  	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
[820](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:821)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[821](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:822)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
[822](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:823)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
[823](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:824)
  	at java.base/java.lang.Thread.run(Thread.java:833)
[824](https://github.com/apache/pulsar/actions/runs/3338223656/jobs/5525935706#step:10:825)
  Caused by: org.apache.bookkeeper.mledger.ManagedLedgerException$BadVersionException: org.apache.pulsar.metadata.api.MetadataStoreException$BadVersionException: org.apache.zookeeper.KeeperException$BadVersionException: KeeperErrorCode = BadVersion for /managed-ledgers/tnx/ns1/persistent/test-txn-topic-test__transaction_pending_ack/__pending_ack_state
```

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStoreTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStoreTest.java
@@ -256,8 +256,8 @@ public class MLPendingAckStoreTest extends TransactionTestBase {
         Assert.assertTrue(mlPendingAckStoreForRead.pendingAckLogIndex.keySet().iterator().next().getEntryId() > 19);
 
         // cleanup.
-        closePendingAckStoreWithRetry(mlPendingAckStoreForWrite, 3);
-        closePendingAckStoreWithRetry(mlPendingAckStoreForRead, 3);
+        closePendingAckStoreWithRetry(mlPendingAckStoreForWrite);
+        closePendingAckStoreWithRetry(mlPendingAckStoreForRead);
     }
 
     /**
@@ -265,19 +265,15 @@ public class MLPendingAckStoreTest extends TransactionTestBase {
      * Because when the cursor close and cursor switch ledger are concurrent executing, the bad version exception is
      * thrown.
      */
-    private void closePendingAckStoreWithRetry(MLPendingAckStore pendingAckStore, int retryTimes){
-        while (retryTimes > 0){
-            retryTimes--;
-            boolean closeSuccessful = true;
+    private void closePendingAckStoreWithRetry(MLPendingAckStore pendingAckStore){
+        Awaitility.await().until(() -> {
             try {
                 pendingAckStore.closeAsync().get();
+                return true;
             } catch (Exception ex){
-                closeSuccessful = false;
+                return false;
             }
-            if (closeSuccessful){
-                break;
-            }
-        }
+        });
     }
 
     /**

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStoreTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStoreTest.java
@@ -256,8 +256,28 @@ public class MLPendingAckStoreTest extends TransactionTestBase {
         Assert.assertTrue(mlPendingAckStoreForRead.pendingAckLogIndex.keySet().iterator().next().getEntryId() > 19);
 
         // cleanup.
-        mlPendingAckStoreForWrite.closeAsync().get();
-        mlPendingAckStoreForRead.closeAsync().get();
+        closePendingAckStoreWithRetry(mlPendingAckStoreForWrite, 3);
+        closePendingAckStoreWithRetry(mlPendingAckStoreForRead, 3);
+    }
+
+    /**
+     * Why should retry?
+     * Because when the cursor close and cursor switch ledger are concurrent executing, the bad version exception is
+     * thrown.
+     */
+    private void closePendingAckStoreWithRetry(MLPendingAckStore pendingAckStore, int retryTimes){
+        while (retryTimes > 0){
+            retryTimes--;
+            boolean closeSuccessful = true;
+            try {
+                pendingAckStore.closeAsync().get();
+            } catch (Exception ex){
+                closeSuccessful = false;
+            }
+            if (closeSuccessful){
+                break;
+            }
+        }
     }
 
     /**

